src/basic/chase-symlinks.c

```diff
@@ -25,23 +26,24 @@ bool unsafe_transition(const struct stat *a, const struct stat *b) {
 }
 
 static int log_unsafe_transition(int a, int b, const char *path, ChaseSymlinksFlags flags) {
-        _cleanup_free_ char *n1 = NULL, *n2 = NULL, *user_a = NULL, *user_b = NULL;
-        struct stat st;
+        _cleanup_free_ char *n1 = NULL, *n2 = NULL;
+        struct stat st, st2;
 
         if (!FLAGS_SET(flags, CHASE_WARN))
                 return -ENOLINK;
 
         (void) fd_get_path(a, &n1);
         (void) fd_get_path(b, &n2);
 
-        if (fstat(a, &st) == 0)
-                user_a = uid_to_name(st.st_uid);
-        if (fstat(b, &st) == 0)
-                user_b = uid_to_name(st.st_uid);
-
-        return log_warning_errno(SYNTHETIC_ERRNO(ENOLINK),
-                                 "Detected unsafe path transition %s (owned by %s) %s %s (owned by %s) during canonicalization of %s.",
-                                 strna(n1), strna(user_a), special_glyph(SPECIAL_GLYPH_ARROW_RIGHT), strna(n2), strna(user_b), path);
+        if (fstat(a, &st) == 0 &&
+            fstat(b, &st2) == 0)
+                return log_warning_errno(SYNTHETIC_ERRNO(ENOLINK),
+                                         "Detected unsafe path transition %s (owned by "UID_FMT") %s %s (owned by "UID_FMT") during canonicalization of %s.",
+                                         strna(n1), st.st_uid, special_glyph(SPECIAL_GLYPH_ARROW_RIGHT), strna(n2), st2.st_uid, path);
+        else
+                return log_warning_errno(SYNTHETIC_ERRNO(ENOLINK),
+                                         "Detected unsafe path transition %s %s %s during canonicalization of %s.",
+                                         strna(n1), special_glyph(SPECIAL_GLYPH_ARROW_RIGHT), strna(n2), path);
 }
```

Comment:
we currently use CHASE_WARN only from dissect and from tmpfiles. Not from PID1 code. Hence this isn't a real issue right now. it is a theoretical one though, which deserves attention.

I think mentioning the actual username is incredibly useful though in tmpfiles, hence I'd like to really keep it there.

so two ideas:

1. just rename CHASE_WARN to CHASE_WARN_WITH_NSS_LOOKUP or so, to make very clear that this will trigger NSS lookups and we hopefully don't accidentally call it from PID1.

2. Add a new flag CHASE_NSS_LOOKUP_OK, and then switch between messages that only include the UID, and those which show the user name. But this might be a bit overkill, given we at least for now do not use CHASE_WARN from PID 1, so these messages would never be triggered.

Reply:
I did a variant of 1., but just added a comment. I think that after all the discussion we had here, we'll remember about the issue, and if not, the comment will be there to remind us.